pkg/controller/architecture/architecture_test.go

```diff
@@ -0,0 +1,133 @@
+
+	// checking if all operations occurred without any issues on the locks
+	if commonMap.Length() != expectedNumberOfNodes {
+		t.Errorf("unxepected number of node information was found: exp=%d act=%d", expectedNumberOfNodes, commonMap.Length())
+	}
+
+}
+
+func TestRead(t *testing.T) {
+	var wg sync.WaitGroup // using wait group to know if task was completed
+	var commonMap = New()
+	var expectedNumberOfNodes = 1000
+	var nodes []*corev1.Node
+
+	for i := 0; i < expectedNumberOfNodes; i++ {
+		node, err := generateNode("node", "amd64", "linux")
+		if err != nil {
+			t.Errorf("unxepected error when generating the node info: %s", err)
+		}
+		err = commonMap.AddNode(node)
+		if err != nil {
+			t.Errorf("unxepected error when appending the node info: %s", err)
+		}
+		nodes = append(nodes, node)
+	}
+
+	for _, node := range nodes {
+		wg.Add(1)
+		go func() {
+			arch, err := commonMap.GetNodeArchitecture(node.Name)
+			if err != nil {
+				t.Errorf("unxepected error when reading node information: %s", err)
+			}
+			if expectedArch, ok := node.Labels["kubernetes.io/arch"]; ok {
+				if arch.Architecture != expectedArch {
+					t.Errorf("unxepected node architecture was found: exp=%q act=%q", expectedArch, arch.Architecture)
+				}
+			}
+			if expectedOS, ok := node.Labels["kubernetes.io/os"]; ok {
+				if arch.OS != expectedOS {
+					t.Errorf("unxepected node architecture was found: exp=%q act=%q", expectedOS, arch.OS)
+				}
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	if commonMap.Length() != expectedNumberOfNodes {
+		t.Errorf("unxepected number of node information was found: exp=%d act=%d", expectedNumberOfNodes, commonMap.Length())
+	}
+
+}
+
+func TestDelete(t *testing.T) {
+	var wg sync.WaitGroup // using wait group to know if task was completed
+	var commonMap = New()
+	var expectedNumberOfNodes = 1000
+	var nodes []*corev1.Node
+
+	for i := 0; i < expectedNumberOfNodes; i++ {
+		node, err := generateNode("node", "amd64", "linux")
+		if err != nil {
+			t.Errorf("unxepected error when generating the node info: %s", err)
+		}
+		err = commonMap.AddNode(node)
+		if err != nil {
+			t.Errorf("unxepected error when appending the node info: %s", err)
+		}
+		nodes = append(nodes, node)
+	}
+
+	for _, node := range nodes {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := commonMap.DeleteNode(node.Name)
+			if err != nil {
+				t.Errorf("unxepected error when deleting the node info: %s", err)
+			}
+		}()
+		wg.Wait()
+	}
+
+	// checking if all operations occurred without any issues on the locks
+	if commonMap.Length() != 0 {
+		t.Errorf("unxepected number of node information was found: exp=%d act=%d", 0, commonMap.Length())
+	}
+
+}
+
+func generateNode(name, arch, os string) (*corev1.Node, error) {
+	suffix, err := uuid.NewRandom()
+	node := &corev1.Node{}
+	node.Name = fmt.Sprintf("%s%d", name, suffix.ID())
+	node.Labels = map[string]string{
+		"kubernetes.io/arch": arch,
+		"kubernetes.io/os":   os,
+	}
+	return node, err
+}
```

Comment:
Not a huge fan of this testing approach.

I think I would like to see a table based approach for each func.

Every test has an input of the nodes that exist, and run the target func on a specific output. We then have what we expect to have as an output.

Reply:
We don't need to test concurrency here.

Reply:
If don't need to test for concurrency should we have a test suite?
since we'll be just testing basic operations on a map

Reply:
Up to you, but it wont hurt